File: apps/ios/Sources/Utils/DataConverter.swift
//
// SmartBLE - Data Converter
//

import Foundation

enum DataFormat {
    case utf8
    case hex
    case base64
}

class DataConverter {
    
    /// 字节数据转十六进制字符串
    static func bytesToHex(_ data: Data, separator: Bool = true) -> String {
        let hex = data.map { String(format: "%02X", $0) }
        return hex.joined(separator: separator ? " " : "")
    }

    /// 十六进制字符串转字节数据
    static func hexToBytes(_ hex: String) -> Data {
        let clean = hex.replacingOccurrences(of: "[^0-9A-Fa-f]", with: "", options: .regularExpression)
        var data = Data()
        var startIndex = clean.startIndex
        while startIndex < clean.endIndex {
            let endIndex = clean.index(startIndex, offsetBy: 2, limitedBy: clean.endIndex) ?? clean.endIndex
            if let byte = UInt8(clean[startIndex..<endIndex], radix: 16) {
                data.append(byte)
            }
            startIndex = endIndex
        }
        return data
    }
    
    /// 字节数据转 UTF-8 字符串
    static func bytesToString(_ data: Data) -> String {
        return String(data: data, encoding: .utf8) ?? bytesToHex(data)
    }
    
    /// 字符串转字节数据
    static func stringToBytes(_ str: String) -> Data {
        return str.data(using: .utf8) ?? Data()
    }
    
    /// 验证是否是合法的十六进制输入
    static func isValidHex(_ hex: String) -> Bool {
        let clean = hex.trimmingCharacters(in: .whitespacesAndNewlines).replacingOccurrences(of: " ", with: "")
        if clean.isEmpty || clean.count % 2 != 0 { return false }
        let regex = try? NSRegularExpression(pattern: "^[0-9A-Fa-f]+$")
        let range = NSRange(location: 0, length: clean.utf16.count)
        return regex?.firstMatch(in: clean, options: [], range: range) != nil
    }
}

File: apps/android/app/src/main/java/com/smartble/core/utils/DataConverter.kt
package com.smartble.core.utils

import java.nio.charset.Charset

object DataConverter {
    /**
     * 将字节数组转换为十六进制字符串
     * @param data 字节数组
     * @param separator 是否使用空格分隔
     */
    fun bytesToHex(data: ByteArray, separator: Boolean = true): String {
        return data.joinToString(if (separator) " " else "") { "%02X".format(it) }
    }

    /**
     * 将十六进制字符串转换为字节数组
     * @param hex 十六进制字符串
     */
    fun hexToBytes(hex: String): ByteArray {
        val clean = hex.replace(Regex("[^0-9A-Fa-f]"), "")
        require(clean.length % 2 == 0) { "Hex 字符串长度必须是偶数" }
        return clean.chunked(2).map { it.toInt(16).toByte() }.toByteArray()
    }

    /**
     * 判断字符串是否是合法的 Hex
     */
    fun isValidHex(hex: String): Boolean {
        val clean = hex.replace(Regex("\\s+"), "")
        if (clean.isEmpty() || clean.length % 2 != 0) return false
        return clean.matches(Regex("^[0-9A-Fa-f]+$"))
    }

    /**
     * 字节数组转字符串
     */
    fun bytesToString(data: ByteArray, charset: Charset = Charsets.UTF_8): String {
        return String(data, charset)
    }

    /**
     * 字符串转字节数组
     */
    fun stringToBytes(str: String, charset: Charset = Charsets.UTF_8): ByteArray {
        return str.toByteArray(charset)
    }
}

// 扩展属性为了向前兼容UI中现有的调用，推荐逐渐替换为直接调用 DataConverter
fun ByteArray.toHexString(separator: Boolean = true): String = DataConverter.bytesToHex(this, separator)
fun String.hexToByteArray(): ByteArray = DataConverter.hexToBytes(this)
